std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,609 @@
+    //             row.a_prime_prime[y][x][limb] = (limb * BITS_PER_LIMB..(limb + 1) * BITS_PER_LIMB)
+    //                 .rev()
+    //                 .fold(F::zero(), |acc, z| {
+    //                     let bit = xor([
+    //                         row.b(x, y, z),
+    //                         andn(row.b((x + 1) % 5, y, z), row.b((x + 2) % 5, y, z)),
+    //                     ]);
+    //                     acc.double() + bit
+    //                 });
+    //         }
+    //     }
+    // }
+
+    let query_a_prime_prime: int, int, int -> int = query |x, y, limb| 
+        utils::fold(
+            16, 
+            |z| 
+                int(eval(b(x, y, (limb + 1) * 16 - 1 - z))) ^ 
+                int(eval(andn(b((x + 1) % 5, y, (limb + 1) * 16 - 1 - z), 
+                b((x + 2) % 5, y, (limb + 1) * 16 - 1 - z)))), 
+            0, 
+            |acc, e| acc * 2 + e
+        );
+
+    query |row| {
+        let _ = array::map_enumerated(a_prime_prime, |i, a_i| {
+            let y = i / 20;
+            let x = (i / 4) % 5;
+            let limb = i % 4;
+
+            provide_value(a_i, row, fe(query_a_prime_prime(x, y, limb)));
+        });
+    };
+
+    // // For the XOR, we split A''[0, 0] to bits.
+    // let mut val = 0; // smaller address correspond to less significant limb
+    // for limb in 0..U64_LIMBS {
+    //     let val_limb = row.a_prime_prime[0][0][limb].as_canonical_u64();
+    //     val |= val_limb << (limb * BITS_PER_LIMB);
+    // }
+    // let val_bits: Vec<bool> = (0..64) // smaller address correspond to less significant bit
+    //     .scan(val, |acc, _| {
+    //         let bit = (*acc & 1) != 0;
+    //         *acc >>= 1;
+    //         Some(bit)
+    //     })
+    //     .collect();
+    // for (i, bit) in row.a_prime_prime_0_0_bits.iter_mut().enumerate() {
+    //     *bit = F::from_bool(val_bits[i]);
+    // }
+
+    query |row| {
+        let _ = array::map_enumerated(a_prime_prime_0_0_bits, |i, a_i| {
+            let limb = i / 16;
+            let bit_in_limb = i % 16;
+
+            provide_value(
+                a_i, 
+                row, 
+                fe((int(eval(a_prime_prime[limb])) >> bit_in_limb) & 0x1)
+            );
+        });
+    };
+
+    // // A''[0, 0] is additionally xor'd with RC.
+    // for limb in 0..U64_LIMBS {
+    //     let rc_lo = rc_value_limb(round, limb);
+    //     row.a_prime_prime_prime_0_0_limbs[limb] =
+    //         F::from_canonical_u16(row.a_prime_prime[0][0][limb].as_canonical_u64() as u16 ^ rc_lo);
+    // }
+
+    let query_a_prime_prime_prime_0_0_limbs: int -> int = query |limb| 
+        int(eval(a_prime_prime[limb])) ^ 
+        ((int(eval(utils::sum(NUM_ROUNDS, |r| expr(RC[r]) * step_flags[r]))) >> (limb * 16)) & 0xffff);
+
+    query |row| {
+        let _ = array::new(4, |limb| {
+            provide_value(a_prime_prime_prime_0_0_limbs[limb], row, fe(query_a_prime_prime_prime_0_0_limbs(limb)));
+        });
+    };
```

Comment:
```suggestion
    let query_a_prime_prime_prime_0_0_limbs: int, int -> int = query |round, limb| 
        int(eval(a_prime_prime[limb])) ^ ((RC[round] >> (limb * 16)) & 0xffff);

    query |row| {
        let _ = array::new(4, |limb| {
            provide_value(a_prime_prime_prime_0_0_limbs[limb], row, fe(query_a_prime_prime_prime_0_0_limbs(row % NUM_ROUNDS, limb)));
        });
    };
```

This is simpler and fixes the BabyBear case!

The problem was that `expr(RC[r])` requires `RC[r]` to fit into the field, because constants in expressions are always field elements.